horovod/run/run.py

```diff
@@ -0,0 +1,398 @@
+
+        # Try ssh 5 times
+        for i in range(SSH_RETRIES):
+            output = six.StringIO()
+            try:
+                exit_code = safe_shell_exec.execute(command,
+                                                    stdout=output,
+                                                    stderr=output)
+                if exit_code == 0:
+                    break
+                else:
+                    output_msg = output.getvalue()
+            finally:
+                output.close()
+        return exit_code, output_msg
+
+    if ssh_port:
+        ssh_port_arg = "-p {ssh_port}".format(ssh_port=ssh_port)
+    else:
+        ssh_port_arg = ""
+
+    ssh_command_format = 'ssh -o StrictHostKeyChecking=no {host} {ssh_port_arg} date'
+    args_list = [[ssh_command_format.format(host=host_address,
+                                            ssh_port_arg=ssh_port_arg)]
+                 for host_address in host_addresses]
+    ssh_exit_codes = \
+        threads.execute_function_multithreaded(exec_command,
+                                               args_list)
+
+    ssh_successful_to_all_hosts = True
+    for index, ssh_status in six.iteritems(ssh_exit_codes):
+        exit_code, output_msg = ssh_status[0], ssh_status[1]
+        if exit_code != 0:
+            print("ssh not successful for host {host}:{msg_output}\n".format(
+                host=host_addresses[index],
+                msg_output=output_msg
+            ))
+
+            ssh_successful_to_all_hosts = False
+    if not ssh_successful_to_all_hosts:
+        exit(1)
+    return True
+
+
+def _launch_task_servers(host_addresses, driver_addresses, num_hosts, tmout,
+                         key, ssh_port=None):
+    """
+    executes the task server and service client task for registration on the
+    hosts.
+    :param host_addresses: list of addresses. for example,
+        ['worker-0','worker-1']
+        ['10.11.11.11', '10.11.11.12']
+    :type host_addresses: list(string)
+    :param driver_addresses: map of interfaces and their address and port for
+    the service. For example:
+        {
+            'lo': [('127.0.0.1', 34588)],
+            'docker0': [('172.122.10.1', 34588)],
+            'eth0': [('11.111.33.73', 34588)]
+        }
+    :type driver_addresses: map
+    :param num_hosts:
+    :type num_hosts: int
+    :param tmout:
+    :type tmout: horovod.spark.util.timeout.Timeout
+    :return:
+    :rtype:
+    """
+
+    def _exec_command(command):
+        host_output = six.StringIO()
+        try:
+            exit_code = safe_shell_exec.execute(command,
+                                                stdout=host_output,
+                                                stderr=host_output)
+            if exit_code != 0:
+                print(
+                    "Launching horovodrun task function was not "
+                    "successful:{host_output}\n"
+                        .format(host_output=host_output.getvalue()))
```

Comment:
Erm, newline should be between ":" and {host_output}, otherwise you have extra newline at the end and no newlines between your message and the output.